osquery/main/shell.cpp

```diff
@@ -83,6 +89,43 @@ int profile(int argc, char *argv[]) {
   return 0;
 }
 
+// readline completion expects strings to be malloced. readline will free them
+// later.
+char *copy_string(const std::string &str) {
+  char *copy = (char *)malloc(str.size() + 1);
+  if (copy == nullptr) {
+    LOG(ERROR)
+        << "Memory allocation failed during shell autocompletion. Exiting!";
+    osquery::Initializer::shutdown(EXIT_FAILURE);
+  }
+  strncpy(copy, str.c_str(), str.size() + 1);
+  return copy;
+}
+
+char *completion_generator(const char *text, int state) {
+  const static std::vector<std::string> tables =
```

Comment:
This does not need to be static given that the perf just needs to be good enough for the interactive use when the user hits `TAB`.